### ISSUE TITLE
ES Backports

### ISSUE THREAD

**Issue by mfussenegger** [crate/crate#9796]
This is a meta issue to track some of the changes in ES that we probably want to backport.
Some of these probably will have a long tail and pull in many other changes.

## On-heap / off-heap

- [x] https://github.com/elastic/elasticsearch/pull/42838 Use reader attributes to control term dict memory useage
- [x] https://github.com/elastic/elasticsearch/pull/52405 Move the terms index of `_id` off-heap

## MetaData storage / ClusterState

- [ ] https://github.com/elastic/elasticsearch/issues/48701 Reduce the number of writes needed for metadata updates
- [ ] https://github.com/elastic/elasticsearch/pull/50144 Use Lucene exclusively for metadata storage
- [ ] https://github.com/elastic/elasticsearch/pull/45007 More logging for slow cluster state application
- [ ] https://github.com/elastic/elasticsearch/pull/47005 Warn on slow metadata persistence

## Engine

- [x] https://github.com/elastic/elasticsearch/pull/48843 Allow realtime get to read from translog
- [x] https://github.com/elastic/elasticsearch/pull/48639 Fix ref count handling in Engine.failEngine
- [ ] https://github.com/elastic/elasticsearch/pull/36038 Set Lucene version upon index creation

## Translog

- [ ] https://github.com/elastic/elasticsearch/commit/41d3eb31946a1f6f894fe1c46f57260e2195704e
- [ ] https://github.com/elastic/elasticsearch/commit/2b8c7c5e11c2d8ea9397bd94e59c5bc7c50c3be4
- [ ] https://github.com/elastic/elasticsearch/commit/6531369f11d159896957b75a7eae0b7da214d01a
- [ ] https://github.com/elastic/elasticsearch/commit/972b81f8a9de5e88106180a5becc5baab39ccdeb
- [ ] https://github.com/elastic/elasticsearch/commit/b034d1e2ef81bcf150ea7d7c115009d2e71545f6 (requires replication/recovery changes)

## Replication

- [ ] https://github.com/elastic/elasticsearch/pull/43205 Advance checkpoints only after persisting ops
- [ ] https://github.com/elastic/elasticsearch/pull/49114 Ignore Lucene index in peer recovery if translog corrupted 
- [ ] https://github.com/elastic/elasticsearch/pull/49746 Replicate write actions before fsyncing them

## Recovery (soft-deletes)

- [ ] https://github.com/elastic/elasticsearch/issues/50775 Make soft-deletes mandatory
- [ ] https://github.com/elastic/elasticsearch/pull/45136 Use index for peer recovery instead of translog
- [ ] https://github.com/elastic/elasticsearch/pull/45473 Ignore translog retention policy if soft-deletes enabled
- [ ] https://github.com/elastic/elasticsearch/commit/09b46c86463a3628dd216e33d23c5fb9495a047c Goodbye and thank you synced flush
- [ ] https://github.com/elastic/elasticsearch/pull/53646 Deprecated support for delaying state recovery pending master nodes
- [ ] https://github.com/elastic/elasticsearch/pull/43787 Make peer recovery clean files step async
- [ ] https://github.com/elastic/elasticsearch/pull/45208 Only retain reasonable history for peer recoveries
- [ ] https://github.com/elastic/elasticsearch/pull/43126 Account soft deletes for committed segments
- [ ] https://github.com/elastic/elasticsearch/pull/43336 Do not use soft-deletes to resolve indexing strategy 
- [ ] https://github.com/elastic/elasticsearch/pull/55353 Retry failed peer recovery due to transient errors
- [ ] https://github.com/elastic/elasticsearch/pull/56072 Avoid copying file chunks in peer covery 

## Snapshots

- [ ] https://github.com/elastic/elasticsearch/pull/51074 Add CoolDown Period to S3 Repository
- [ ] https://github.com/elastic/elasticsearch/pull/49862 Cleanup Old index-N Blobs in Repository Cleanup
- [ ] https://github.com/elastic/elasticsearch/pull/55918 Remove Snapshot INIT Step
- [ ] https://github.com/elastic/elasticsearch/pull/56669 More Efficient Snapshot State Handling 
- [ ] https://github.com/elastic/elasticsearch/pull/58214

## Other 

- [ ] https://github.com/elastic/elasticsearch/pull/48804 Optimize sort on long field
- [ ] https://github.com/elastic/elasticsearch/pull/44433 Defer reroute when starting shards 
- [ ] https://github.com/elastic/elasticsearch/pull/44485 Improve handshake failure messages
- [ ] https://github.com/elastic/elasticsearch/pull/44619 Check shard limit after applying index templates

**Comment by mfussenegger** [comment]
Closing this in favour of the es-backports file added in https://github.com/crate/crate/pull/10299
